### DISCUSSION TITLE
grafo con archi di peso 1 o 2

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#32] [Esercizi A-L]
Dato un grafo $G = (V, E)$, data la funzione $\omega : E \to \set{1, 2}$ che associa ad ogni arco $e \in E$ un peso (1 o 2), definiamo la distanza fra due vertici $x, y \in V$ come il cammino con il peso totale minimo.

Scrivere lo pseudocodice di un algoritmo che trova la distanza fra due vertici $x, y$ nel grafo pesato, in $O(n + m)$

**Comment by interestingmaneuver** [comment]
Possiamo adattare Dijkstra eliminando il costo della coda di priorità perché abbiamo solo due valori:

```python
from math import inf
def dijkstra_binary(graph, x, y): # graph come liste di adiacenza
    ones = [x] # anche se non è 1 l'importante è che venga scelta per prima
    twos = []
    unreach = [i for i in range(len(graph)) if i != x]
    distances = [inf] * len(graph)
    distances[x] = 0

    while ones or twos or unreach:
        if ones:
            u = ones.pop()
        elif twos:
            u = twos.pop()
        else:
            u = unreach.pop()
        for (neigh, costo) in graph[u]:
            new_dist = distances[u] + costo
            if new_dist < distances[neigh]:
                distances[neigh] = new_dist
                if costo == 1:
                    ones.append(neigh)
                else:
                    twos.append(neigh)
    return distances[y]

# Esempio:
#       2
# (0) ------ (1)
#  \          /
# 2 \        / 1
#   (2) --- (3)
#        2
# Risultato = 3
g = [
    [(1,2),(2,2)],
    [(3,1),(0,2)],
    [(3,2),(0,2)],
    [(1,1),(2,2)]
]
print(dijkstra_binary(g, 0, 3))

```

**Reply by IonutCicio** [reply]
Dijkstra è magico, ma la challenge _(che non ho specificato)_ è quella di farlo senza Dijkstra, usando un normale BFS, e sfruttando il fatto che il peso può essere o 1 o 2. Di fatto, Dijkstra non ha complessità $O(n + m)$

**Comment by Frisbon** [comment]
### Usando il BFS
![image_2024-06-15_14-17-56](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/5268b220-59fc-483a-922b-d7aac5f35306)

### Esempio
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/18241783-e8cf-4f5c-af73-a81763568dda)
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/6bc6247f-73b3-4c34-84d0-7f9bb52039a3)
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/c77302da-8684-420b-9b2d-3da1fb82c6c3)

**Reply by interestingmaneuver** [reply]
Il tuo algoritmo non sembra poter terminare, difatti a C vengono sempre aggiunti elementi (suppongo tu assumi che ogni nodo non può essere visitato due volte).

Il problema però è che la "comodità per l'esempio" ti inganna, nel tuo caso funziona perché hai scelto quell'ordine preciso. Quando aggiorni `D[v]`, non sai se `D[u]` sia effettivamente il modo meno costoso di raggiungere u, perché non sei sicuro di aver visto già tutti i percorsi che portano ad u.

Puoi anche accorgerti che non hai fatto uso dell'ipotesi che gli archi possono avere solo 2 costi.

**Comment by CasuFrost** [comment]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/97389762/22867732-357c-489c-8bc3-90e216334054)

**Comment by niccolo-fato** [comment]
![photo_2024-07-16_17-01-26](https://github.com/user-attachments/assets/10669899-ce9f-4b38-815f-f2f4dfbb6f39)
